src/Elastic.Apm.AspNetCore/DiagnosticListener/AspNetCoreDiagnosticListener.cs

```diff
@@ -28,7 +28,7 @@ public AspNetCoreDiagnosticListener(IApmAgent agent)
 
 		public void OnError(Exception error) { }
 
-		public void OnNext(KeyValuePair<string, object> kv)
+		public async void OnNext(KeyValuePair<string, object> kv)
```

Comment:
This is questionable. We need to call an async method here. 

Here is what I think:
This is a fire and forget call, meaning the caller will move on. 
2 things can happen: 
- this `OnNext` method finishes before the transaction is finished -> all good, we captured the body. 
- the transaction finishes before we set `Transaction.Request.Body` and in that case the body won't be captured. This is not cool. 

Alternatively we can revert back to async reading in this case, but that would be very unfortunate, because with that we again risk thread starvation. 

Another option would be to mark the transaction here and not do the reading and have a check in `ApmMiddleware.InvokeAsync` where we can call `await CollectRequestInfoAsync`. 

Happy to hear opinions.

Reply:
Yeah, this was a problem. I moved the reading of the body to the `ApmMiddleware`. The CI fails mentioned [here](https://github.com/elastic/apm-agent-dotnet/pull/608#issuecomment-555076734) were caused by this.